numba/targets/arraymath.py

```diff
@@ -1450,3 +1460,158 @@ def impl(arg):
 
 generate_xinfo(np.finfo, finfo, _finfo_supported)
 generate_xinfo(np.iinfo, iinfo, _iinfo_supported)
+
+def _get_inner_prod(dta, dtb):
+    # gets an inner product implementation, if both types are float then
+    # BLAS is used else a local function
+
+    @register_jitable
+    def _innerprod(a, b):
+        acc = 0
+        for i in range(len(a)):
+            acc = acc + a[i] * b[i]
+        return acc
+
+    # no BLAS... use local function regardless
+    if not _HAVE_BLAS:
+        return _innerprod
+
+    flty = types.real_domain | types.complex_domain
+    floats = dta in flty and dtb in flty
+    if not floats:
+        return _innerprod
+    else:
+        a_dt = as_dtype(dta)
+        b_dt = as_dtype(dtb)
+        dt = np.promote_types(a_dt, b_dt)
+
+        @register_jitable
+        def _dot_wrap(a, b):
+            return np.dot(a.astype(dt), b.astype(dt))
+        return _dot_wrap
+
+def _assert_1d(a, func_name):
+    if isinstance(a, types.Array):
+        if not a.ndim <= 1:
+            raise TypingError("%s() only supported on 1D arrays " % func_name)
+
+def _np_correlate_core(ap1, ap2, mode, direction):
+    pass
+
+@overload(_np_correlate_core)
+def _np_correlate_core_impl(ap1, ap2, mode, direction):
+    a_dt = as_dtype(ap1.dtype)
+    b_dt = as_dtype(ap2.dtype)
+    dt = np.promote_types(a_dt, b_dt)
+    innerprod = _get_inner_prod(ap1.dtype, ap2.dtype)
+
+    def impl(ap1, ap2, mode, direction):
+        # Implementation loosely based on `_pyarray_correlate` from
+        # https://github.com/numpy/numpy/blob/3bce2be74f228684ca2895ad02b63953f37e2a9d/numpy/core/src/multiarray/multiarraymodule.c#L1191
+        # For "Mode":
+        # Convolve uses 'full' by default, this is denoted by the number 2
+        # Correlate uses 'valid' by default, this is denoted by the number 0
+        # For "direction", +1 to write the return values out in order 0->N
+        # -1 to write them out N->0.
+
+        if not (mode == 0 or mode == 2):
```

Comment:
Can the `0` and `2` be named as sth like `CORR_FULL=0` or use IntEnum for better readability?

Reply:
Also, I curious to know why `1` is skipped.

Reply:
This function is shared by correlate and convolve, in NumPy these functions have a `mode` kwarg (which is not in this PR) and the default mode is different in each function. The `mode` ends up in `C` code as having value `0, 1, 2` based on the string passed to `mode` (one of `'valid', 'same', 'full'`), this function does the conversion https://github.com/numpy/numpy/blob/v1.14.0/numpy/core/numeric.py#L862-L870. To try and reduce future confusion, I've kept the magic numbers the same as in the NumPy impl (made a note here https://github.com/numba/numba/pull/2777/files#diff-4b9c6924f184d92f076cc50b84bcc442R1512). Good idea RE the `IntEnum`, will add one with members based on the strings for `mode` to enumerate the magic numbers.

Reply:
Oh and `1` is skipped as the mode `same` is not in use at present. :)